src/Symfony/Component/EventDispatcher/EventDispatcher.php

```diff
@@ -46,21 +44,23 @@ public function dispatch($eventName, Event $event = null)
         $event->setDispatcher($this);
         $event->setName($eventName);
 
-        if (!isset($this->listeners[$eventName])) {
-            return $event;
+        if ($listeners = $this->getListeners($eventName)) {
+            $this->doDispatch($listeners, $eventName, $event);
         }
 
-        $this->doDispatch($this->getListeners($eventName), $eventName, $event);
-
         return $event;
     }
 
     /**
-     * @see EventDispatcherInterface::getListeners()
+     * {@inheritdoc}
      */
     public function getListeners($eventName = null)
     {
         if (null !== $eventName) {
+            if (!isset($this->listeners[$eventName])) {
+                return array();
```

Comment:
without it, the `sorted` property would be initialized even for non-existing events. Getters should not leak infinite memory.